src/language_features/semantic_tokens.rs

```diff
@@ -65,18 +66,53 @@ pub fn tokens_response(meta: EditorMeta, tokens: SemanticTokensResult, ctx: &mut
                     end: Position::new(line, start + length),
                 };
                 let range = lsp_range_to_kakoune(&range, &document.text, ctx.offset_encoding);
-                let token = &legend.token_types[token_type as usize];
-                (0..32)
+                let token_name = legend.token_types[token_type as usize].as_str();
+                let token_modifiers: Vec<&SemanticTokenModifier> = (0..32)
+                    // Find bits in the mask that equal `1`
                     .filter(|bit| ((token_modifiers_bitset >> bit) & 1u32) == 1u32)
+                    // Map bits to modifiers
                     .map(|bit| &legend.token_modifiers[bit as usize])
-                    .filter_map(|token| ctx.config.semantic_token_modifiers.get(token.as_str()))
-                    .chain(ctx.config.semantic_tokens.get(token.as_str()))
-                    .next()
-                    .map(|face| format!("{}|{}", range, face))
+                    .collect();
+
+                let mut candidates = ctx
+                    .config
+                    .semantic_tokens
+                    .iter()
+                    .filter_map(|token_config| {
```

Comment:
Clippy likes to complain about "`filter` followed by `map`" actually.
Moving it to `sort_by_cached_key` could make sense, though.

But I don't think it would have much of an effect, so I'm fine with any option.

Reply:
In this instance clippy doesn't complain about filter+map.
Looks like there were some recent changes that probably made it more specific.
I imagine it should only apply if the closures in filter+map have some duplicate code.

The advantage of smaller filters is smaller scope, though I agree that any of these options is fine.

However, we can just use `Iterator::max_by_key` instead:

```rust
                let candidates = ctx.config.semantic_tokens.iter().filter(|token_config| {
                    token_name == token_config.name &&
                        // All the config's modifiers must exist on the token for this
                        // config to match.
                        token_config
                            .modifiers
                            .iter()
                            .all(|modifier| token_modifiers.contains(&modifier))
                });
                // Not all the token's modifiers must exist in the config.  Therefore, we use
                // the token that matches the most configured modifiers.
                let best = candidates.max_by_key(|token_config| {
                    token_modifiers
                        .iter()
                        .filter(|modifier| token_config.modifiers.contains(modifier))
                        .count()
                });
                best.map(|token_config| format!("{}|{}", range, token_config.face))
```

Reply:
Ah, didn't know about `max_by_key`. Neat.

Reply:
Hold on, did this change the behavior?
Previously we always used the last token-config if they had equal number of matches (because Vec::sort is a stable sort).
I think that's no longer guaranteed (didn't test it though).

So maybe https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold would be the "right" way to implement this..

---

In general I'm not sure what's the best UI for this prioritization feature.
"first wins" / "last wins", or maybe adding an explicit "priority" field. I'm interested in how other editors do it.

Reply:
> Hold on, did this change the behavior?

It didn't:

> If several elements are equally maximum, the last element is returned
> <https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.max_by_key>

---

I don't think there's much of a difference in UX between the three options.

One additional option would be to only match against "token + identical modifiers" and "token + no modifiers". That would be easier to understand, but quite annoying to configure, since some servers will regularly apply 3-5 modifiers, and there are no tools to visualize that (afaik).